File: node/packages/aws-lambda-sdk/trace-spans/aws-lambda.js
'use strict';

const TraceSpan = require('@serverless/sdk/lib/trace-span');

const immutableTags = {
  'aws.lambda.name': process.env.AWS_LAMBDA_FUNCTION_NAME,
  'aws.lambda.version': process.env.AWS_LAMBDA_FUNCTION_VERSION,
};

const arch = (() => {
  switch (process.arch) {
    case 'x64':
      return 'x86_64';
    case 'arm64':
      return 'arm64';
    default:
      console.warn({
        source: 'serverlessSdk',
        message: `Unrecognized architecture: "${process.arch}"\n`,
        code: 'UNRECOGNIZED_ARCH',
      });
      return null;
  }
})();
if (arch) immutableTags['aws.lambda.arch'] = arch;

const awsLambdaSpan = new TraceSpan('aws.lambda', {
  startTime: EvalError.$serverlessAwsLambdaInitializationStartTime,
  immediateDescendants: ['aws.lambda.initialization'],
  tags: immutableTags,
});

if (process.env.AWS_LAMBDA_INITIALIZATION_TYPE === 'on-demand') {
  awsLambdaSpan.tags.set('aws.lambda.is_coldstart', true);
}

awsLambdaSpan.tags.reset = function () {
  this.clear();
  this.setMany(immutableTags);
};

module.exports = awsLambdaSpan;

File: python/packages/aws-lambda-sdk/serverless_aws_lambda_sdk/trace_spans/aws_lambda.py
from typing_extensions import Final
import os
import platform
import logging

from sls_sdk.lib.trace import TraceSpan
from sls_sdk.lib.timing import _DIFF

__all__ = [
    "aws_lambda_span",
]


IMMUTABLE_TAGS: Final[dict] = {
    "aws.lambda.name": os.environ.get("AWS_LAMBDA_FUNCTION_NAME"),
    "aws.lambda.version": os.environ.get("AWS_LAMBDA_FUNCTION_VERSION"),
}


def _arch():
    arch = platform.machine()
    if arch == "x64" or arch == "x86_64":
        return "x86_64"
    elif arch == "aarch64" or arch == "arm64":
        return "arm64"
    else:
        logging.warning(
            {
                "source": "serverlessSdk",
                "message": f'Unrecognized architecture: "{arch}"',
            }
        )
    return None


arch = _arch()
if arch:
    IMMUTABLE_TAGS["aws.lambda.arch"] = arch


aws_lambda_span = TraceSpan(
    "aws.lambda",
    start_time=int(os.environ.get("_SLS_PROCESS_START_TIME", 0)) - _DIFF,
    immediate_descendants=["aws.lambda.initialization"],
    tags=IMMUTABLE_TAGS,
)

if os.environ.get("AWS_LAMBDA_INITIALIZATION_TYPE") == "on-demand":
    aws_lambda_span.tags["aws.lambda.is_coldstart"] = True


def _clear(self: TraceSpan):
    self.tags.clear()
    self.tags.update(IMMUTABLE_TAGS)
    self.sub_spans.clear()


TraceSpan.clear = _clear
